src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+        switch(thread->second.state) {
+          case TASK_BUFN:
+          //fall through
+          case TASK_BUFN_THROW:
+          //fall through
+          case TASK_BUFN_WAIT:
+            transition(thread->second, thread_state::TASK_RUNNING);
+            task_wake_condition.notify_all();
+            break;
+          default:
+            break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Internal implementation that removes a threads association with a task/shuffle.
+   * returns true if the thread that ended was a normally running task thread.
+   * This should be used to decide if wake_up_threads_after_task_finishes is called or not.
+   */
+  bool remove_thread_association(long thread_id, std::unique_lock<std::mutex> & lock) {
+    bool ret = false;
+    auto threads_at = threads.find(thread_id);
+    if (threads_at != threads.end()) {
+      auto task_id = threads_at->second.task_id;
+      if (task_id >= 0) {
+        auto task_at = task_to_threads.find(task_id);
+        if (task_at != task_to_threads.end()) {
+          task_at->second.erase(thread_id);
+        }
+      }
+
+      switch (threads_at->second.state) {
+        case TASK_BLOCKED:
+        // fall through
+        case TASK_BUFN:
+          transition(threads_at->second, thread_state::TASK_REMOVE_THROW);
+          task_wake_condition.notify_all();
+          break;
+        case SHUFFLE_BLOCKED:
+          transition(threads_at->second, thread_state::SHUFFLE_REMOVE_THROW);
+          task_wake_condition.notify_all();
+          break;
+        case TASK_RUNNING:
+          ret = true;
+          // fall through;
+        default:
+          log_transition(thread_id, threads_at->second.task_id, threads_at->second.state, thread_state::UNKNOWN);
+          threads.erase(threads_at);
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Called prior to processing an alloc attempt. This will throw any injected exception and
+   * wait until the thread is ready to actually do/retry the allocation. That blocking API may
+   * throw other exceptions if rolling back or splitting the input is considered needed.
+   */
+  void pre_alloc(long thread_id) {
+    std::unique_lock<std::mutex> lock(state_mutex);
+
+    auto thread = threads.find(thread_id);
+    if (thread != threads.end()) {
+      if (thread->second.retry_oom_injected > 0) {
+        thread->second.retry_oom_injected--;
+        throw_java_exception(RETRY_OOM_CLASS, "injected RetryOOM");
+      }
+
+      if (thread->second.cudf_exception_injected > 0) {
+        thread->second.cudf_exception_injected--;
+        throw_java_exception(CUDF_EXCEPTION_CLASS, "injected CudfException");
+      }
+
+      if (thread->second.split_and_retry_oom_injected > 0) {
+        thread->second.split_and_retry_oom_injected--;
+        throw_java_exception(SPLIT_AND_RETRY_OOM_CLASS, "injected SplitAndRetryOOM");
+      }
+
+      block_thread_until_ready(lock);
```

Comment:
nit, we could pass `thread_id` here... save us a lookup and check against `threads.end`  `block_thread_until_ready`